### DISCUSSION TITLE
Why is directx no longer available as a renderer ?

### DISCUSSION THREAD

**Discussion by ArnoldUK** [86Box/86Box#2390] [General]
I've been using old version 2.07 for quite a while and since upgrading to version 3.0 the renderers are all opengl and vulkan. using any opengl renderer causes very high CPU usage even for basic 8mhz machine. The directx renderers did not cause high cpu usage and were much more cleaner video display. Just wondered why the switch from directx to opengl ?

**Comment by OBattler** [comment]
I did push for a Direct3D renderer but the person who was working on the port, namely Cacodemon, insisted it was impossible. I personally don't believe it's impossible (multiple things he claimed were impossible, later turned out not to be), but good luck convincing him.

**Reply by OBattler** [reply]
But, a lot of the increased CPU uage is also because of QT itself. At this point, I'm tempted to begin offering Win32 builds alongside QT builds.

**Reply by ArnoldUK** [reply]
Yeah, I agree, the QT build is heavy on the CPU under Windows but I guess the MACOS and Linux side it's an improvement over directx Win32 builds. I've dropped back to build v2.07 as it jsut appears more stable under Windows. Still the 3.0 build is a step forward apart from the renderer now being only opengl and I don't know anything about Vulkan

**Reply by OBattler** [reply]
There is a Direct3D render in 3.0 - SDL (Hardware). That default to Direct3D on Windows. That continued to exist until we went to QT in, I believe, 3.3?

**Reply by ArnoldUK** [reply]
Oh, thanks, I had no idea any of the 3.0 builds still supported D3D hardware. I think I download the wrong build 3.5.
Going to try build 3.0 or 3.3 and see for myself. Cheers! On a side note, I've jumped ship from virtualbox and I'm now using 86Box more for PC emulation. VirtualBox has become heavily bloated and is pushing more to modern pc systems emulations leaving the classic systems with very poor emulation and slow video graphic support. DosBox is great also but the base DOS system that it uses is incompatible with many DOS applications and the machine speed control is, well, out of control :)

**Reply by OBattler** [reply]
On 3.5, did you try the Software render? I was just told that that uses DirectX on QT.

**Reply by ArnoldUK** [reply]
Yes, both QT (software) and QT (OpenGL) casued my cpu (Core I5 2.4Ghz) to be pegged out on both cores running a basic 8086 8Mhz or 486 system. Vulkan renderer is alot less cpu and seems to be more stable without some of the glithes with opengl.

**Reply by OBattler** [reply]
What is your host hardware?

**Reply by ArnoldUK** [reply]
Host System: Toshiba S50 Laptop Core I5 2.4Ghz 8G RAM with Dual Video Intel and AMD Radeon graphics.
OS: Windows 10 with updates.

Build V2.07 is the only build with DirectDraw and Direct3D renderers.
Build V3.0 to V3.2 use SDL software and hardware renderers and also OpenGL 3.0.
Build V3.3 to V3.5 now use QT with OpenGL and Vulkan renderers.

I can def confirm that build V2.07 with Direct3D 9 renderer uses less than 10% cpu. Build V3.2 with SDL (Software) renderer appears to be the best of the V3.0 builds but still pegs the cpu upto 50%. Using SDL (Hardware) or any OpenGL renderer pegs my cpu over 80% in V3.3 builds and later. This is running just a simple emulated 486DX-33 system which is idle at DOS command prompt.

**Reply by OBattler** [reply]
- ArnoldUK: What Core i5 model specifically?

**Reply by ArnoldUK** [reply]
Core i5-4210U
I'm only using standard 32Bit builds and not cpu optimized builds.

**Reply by ArnoldUK** [reply]
Well I can emulate your Core i5-4670K with my 60 Core Intel i9-12900K becuase I just can

**Comment by dhrdlicka** [comment]
What GPU do you have on the host? Are your graphic drivers up to date?

**Reply by ArnoldUK** [reply]
Laptop with discrete/hybrid Intel HD 4400 and Radeon R7 M260. I dont think it's video driver issue as nothing is graphical intensive. I had similar issues with VirtualBox when that startd using QT.

**Reply by dhrdlicka** [reply]
The thing is that VirtualBox hasn't actually made any such switch to Qt at all, they've been using it from the start. Do you remember when you started to experience these issues with VirtualBox?

**Reply by ArnoldUK** [reply]
I'm talking about VirtualBox from 10 years ago, versions 1-3 were the best versions for good old DOS and Windows 31/95 systems. After version 4.0 became heavily bloated and started to target enterprise market trying to compete with vmware and ms hyper-v azure etc.. You go to a VirtualBox forum now and talk about problems installing DOS and Win16 systems and you are fobbed off and told to go use DosBox.

**Reply by ts-korhonen** [reply]
> I dont think it's video driver issue as nothing is graphical intensive

Please try picking up the latest drivers. I've seen it make a ton of difference. With HD4000 + Win10 stock drivers opengl was unusable; glitches and framerate was in the slideshow region. With latest drivers from Intel it worked just fine.

**Reply by ArnoldUK** [reply]
I dont have any issues with opengl or any games that use it. The problem is with QT as OBattler pointed out.

**Reply by OBattler** [reply]
Just to make sure - does the latest 86Box with Win32 (should be 3.2) run fine when using the SDL OpenGL and OpenGL 3.0 renderers?

**Reply by ArnoldUK** [reply]
Yes, build 3.2 runs ok just slightly more cpu usage than 2.x Builds. The 3.5 build hammers my CPU in all renderers apart from vulkan. The 3.0 builds also fix the corrupt mouse with ATI Mach64 Video.

I've also found a fix for the slow jerky mouse issues when using the Logitech/MS Bus mouse.
The problem is with the default Hz setting. Changing the value to non-timed (original) gives very smooth mouse control and very good response.

**Reply by OBattler** [reply]
- ArnoldUK: Try the latest 86Box build, that has a Direct3D renderer.

**Reply by ArnoldUK** [reply]
> well that explains alot. I'm starting to think about double checking task manager on the host and the speed history on the guest.

This isn't a problem with the host as I said that the issue does not exist with v2.x builds running on the exact same guest system and the exact same processes running on the host.

**Reply by ArnoldUK** [reply]
> * ArnoldUK: Try the latest 86Box build, that has a Direct3D renderer.

Yes I am using build v2.10 with D3D Hardware renderer. Works great!

**Reply by OBattler** [reply]
I'm talking about the latest 86Box build, as we readded the Direct3D render.

**Reply by ArnoldUK** [reply]
Version 3.5 is the only build I can download but that does nto have Direct3D renderer, it's OpenGL. Is there a build newer than this ?

**Reply by ArnoldUK** [reply]
Ok, I will give it a try. Thanks!

**Reply by ArnoldUK** [reply]
Just tried latest Jenkins V3.5 build #3803 which does have Direct3D 9 renderer but the video image quality is not very good, looks like a very compressed jpeg image. The high CPU usage also is still there.

**Reply by ArnoldUK** [reply]
Is that a joke ? I mean the emulated display quality looks like a compressed jpeg image, not crisp fonts or graphics.

**Reply by Cacodemon345** [reply]
Are you using the Linear video filter?

**Reply by DDXofficial** [reply]
> Just tried latest Jenkins V3.5 build #3803 which does have Direct3D 9 renderer but the video image quality is not very good, looks like a very compressed jpeg image. The high CPU usage also is still there.

This is particularly strange. Given that this is a graphics-related issue, could you upload a screenshot of the degraded image quality to this thread? Take the screenshot from within the host OS (i.e. outside 86Box) as screenshots from 86Box itself won't be affected.

**Reply by ArnoldUK** [reply]
I've uploaded the screenshots to my Github page https://github.com/ArnoldUK/86Box showing D3D9 and QT OpenGL renderer differences. You can clearly see the D3D9 rendering quality is very poor. Changing the video filter to linear or nearest makes no difference. I have no idea how to upload images to the discussion posts.

The emulated guest machine is as follows:
Socket 5 - Gigabyte i430NX - Intel Pentium 90 - 32MB Memory
PCI ATI Mach64GX 4MB
ISA Soundblaster Pro 2

**Reply by Cacodemon345** [reply]
Is your DPI scaling setting anything but 100%?

**Reply by DDXofficial** [reply]
> I have no idea how to upload images to the discussion posts.

To my knowledge, it's fairly simple - you can drag and drop/copy and paste them to this text box on GitHub. 

That said, I have looked at the screenshots and I do have a clearer understanding of the situation at hand. @Cacodemon345 is currently theorizing that your screen/DPI scaling being higher than 100% may be the issue, hence him asking you the question above.

**Reply by ArnoldUK** [reply]
> 🤔 very compressed jpeg image? No, it's just your CPU

Don't say that, I'm beginning to think my pc is really screwed up now.

**Reply by ArnoldUK** [reply]
OK thanks DDXOfficial, that was too simple, that's why I found it hard to do :0
I've tried every setting under the sun, the DPI scaling is not changed and I am not using HiDPI scaling.
Although when I do enable HiDPI scaling I get similar garbled image to the screenshot but not as bad in OpenGL.

**Reply by OBattler** [reply]
Try the attached binary that's compiled with latest code but using the Win32 UI and see how well it performs: [86Box.zip](https://github.com/86Box/86Box/files/8983684/86Box.zip) .

**Reply by ArnoldUK** [reply]
Thanks for the updated build but I do not see much of difference in cpu usage than the original V3.5, also the D3D renderer is missing. I tested all renderers. Thanks for the update, much appreciated anyway.

**Reply by OBattler** [reply]
The D3D renderer is there - it's called SDL (Hardware).

But this shows us that 86Box itself uses more CPU than it used to. At this point, I wonder what could have caused that.

**Reply by OBattler** [reply]
OK, try this new binary, compiled with OpenAL instead of XAudio2, seems to lower CPU usage on my end: [86Box.zip](https://github.com/86Box/86Box/files/8984795/86Box.zip) .

**Reply by OBattler** [reply]
OK, I don't think it was XAudio2. It was a piece of code in the main thread, the new build is compiling, see if that build still uses a lot of CPU (the usage is significantly reduced on my end).

**Reply by ArnoldUK** [reply]
I've just tried the latest builds #3805 on Jenkins, both compiler versions. The result is worse than the current build on github.
The sound also stutters in Windows guests. The D3D renderer is improved graphic quality but still not as good as the V2.x builds. I beginning think this is an issue with my PC setup as you have tried so much to help and no one else appears to have an issue with the 3.5 build in Windows. THe 3.5 build is still usable, but I will jsut have to live with sound of my CPU fan constantly running with it.

**Reply by OBattler** [reply]
Did you try my build with OpenAL instead of XAudio2? Just to see if that has any effect.

**Reply by ArnoldUK** [reply]
Yes, I tried both of the zip files but no luck with those either. The Jenkins build caused the guest to become very laggy as the cpu % in the tool bar would not go any higher than 50%. The other builds I can get full 100% emulator speed so no lag and the audio does not stutter.

**Reply by OBattler** [reply]
Are you downloading the correct Jenkins build? Ie. not the Debug one as that one is slower. Specifically, make sure you're downloading this one: https://ci.86box.net/job/86Box/lastSuccessfulBuild/artifact/Old%20Recompiler%20(recommended)/Windows%20-%20x86%20(32-bit)/86Box-Windows-32-b3806.zip .

**Reply by ArnoldUK** [reply]
Yes, I did try that version with old recompiler, the build is very slow and want go beyond 50% in the toolbar cpu which makes it very laggy. The Direct3D 9 renderer is much clearer but still a little smudgey, like some kind of filter or interpolation, but I have filter set to Linear as it is much cripser image.
Edit. Here are the screenshots:
https://github.com/ArnoldUK/86Box/blob/main/86Box-Win32-V35-B3806-D3D9-1.png
https://github.com/ArnoldUK/86Box/blob/main/86Box-Win32-V35-B3806-D3D9-2.png

Out of curisoity, what exactly is Jenkins build ? reminds of the classic Ask Jeeves search engine :)

**Reply by OBattler** [reply]
OK, try 86Box v2.07 with the SDL (Software) renderer, which I believe was already present, and see if you get increased CPU usage.

**Reply by ArnoldUK** [reply]
I'm currently using v2.07 b2008 which has the most stable and fast emulation with full 100% emulation speed and less than 10% cpu on my host cpu. The midi out through an emualted MT-32 is all clear and smooth without any stuttering. Both SDL and D3D9 renderers work perfectly well (DirectDraw stutters a little though, but I expect that).

Now having downloaded so may builds of v3.x, I have found v3.2 b3600 to be the best of all of them, although host cpu usage is close to 50% and audio has some skips and stutters. All other v3.x builds including the Jenkins builds are nothing close to performance of the v2.07 and v2.10 release.

You can try this out for yourself and see the clear differences in performance and responsiveness of the emualtion in version v2.07. From what I've gathered so far, seems to me something drastically changed within the v3.x release and I guess it is the combination of QT interface and OpenGL renderers. I also noticed that v2.x builds do not have Filter methods or HiDPI scaling options, maybe that is the change in v3.x that is causing the performance drops. ? I am specfically talking about the Win32/64 builds and this may not apply to Linux or other builds.

**Reply by OBattler** [reply]
Can I see your 86box.cfg file?

**Reply by ArnoldUK** [reply]
It's here on my Github page. I dont know how to use or create folders in Github as I would also upload the ROM nvr folder.
https://github.com/ArnoldUK/86Box/blob/main/86box.cfg

**Reply by ArnoldUK** [reply]
> which one is the other build? lesser than the recent?

The 3.x versions on github releases page, not the Jenkins builds or the zip files OBattler linked.

**Comment by Fenix770** [comment]
I have that same CPU i5 4210U, but with an Nvidia 820M instead of AMD/ATI graphics card. I noticed the same slowdowns. On PCem I can emulate a Cyrix 5x86/90, at 100% on Windows 95,  but with OpenGL 3.0 only a 486 DX/2 50. Surface/software option uses much more CPU than OpenGL 3.0. 
It think , however, I'll need a newer PC before the end of support for Windows 10. 
The machine is still usable with an SDD and 12gb of RAM.

**Reply by OBattler** [reply]
IMHO, comparisons with PCem should be done cautiously, because 86Box compared to PCem already loses performance due to using the old recompiler instead of the new one, and parts having been rewritten for more accuracy.

**Reply by ArnoldUK** [reply]
Fennix770, same here and I use PCem also with almost no CPU load with an emulated pentium. But 86Box is a fork of PCem so I don't understand why the move to OpenGL. Maybe it's a licence issue with directx and not being able to distribute directx dll's ? or maybe directx is difficult to work with on linux/mac os...no idea really.

**Reply by OBattler** [reply]
It's because it's not easy to use DirectX with QT, or to use SDL2 with QT for that matter.

But, since QT Software has been observed to increase CPU load as well, I'm not really sure how it's a renderer issue. It seems more like a QT issue to me, which is why I am considering to begin providing Win32 builds again alongside QT builds, at least until someone finds a way to do a DirectX renderer on QT and fix QT performance.

**Reply by ArnoldUK** [reply]
That would be great for a pure Win32 build. I'm browsing through the code at the moment just to see how everything functions. May even have a go at doing a build myself and porting it to Visual Studio.
